velox/functions/prestosql/TransformKeys.cpp

```diff
@@ -0,0 +1,124 @@
+  bool isDefaultNullBehavior() const override {
+    // transform_keys is null preserving for the map. But
+    // since an expr tree with a lambda depends on all named fields, including
+    // captures, a null in a capture does not automatically make a
+    // null result.
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    VELOX_CHECK_EQ(args.size(), 2);
+
+    // Flatten input map.
+    exec::LocalDecodedVector mapDecoder(context, *args[0], rows);
+    auto& decodedMap = *mapDecoder.get();
+
+    auto flatMap = flattenMap(rows, args[0], decodedMap);
+
+    std::vector<VectorPtr> lambdaArgs = {
+        flatMap->mapKeys(), flatMap->mapValues()};
+    auto numKeys = flatMap->mapKeys()->size();
+
+    VectorPtr transformedKeys;
+
+    // Loop over lambda functions and apply these to keys of the map.
+    // In most cases there will be only one function and the loop will run once.
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+    while (auto entry = it.next()) {
+      auto keyRows =
+          toElementRows<MapVector>(numKeys, *entry.rows, flatMap.get());
+      auto wrapCapture = toWrapCapture<MapVector>(
+          numKeys, entry.callable, *entry.rows, flatMap);
+
+      entry.callable->apply(
+          keyRows, wrapCapture, context, lambdaArgs, &transformedKeys);
+    }
+
+    auto localResult = std::make_shared<MapVector>(
+        flatMap->pool(),
+        outputType,
+        flatMap->nulls(),
+        flatMap->size(),
+        flatMap->offsets(),
+        flatMap->sizes(),
+        transformedKeys,
+        flatMap->mapValues());
+
+    checkDuplicateKeys(localResult, rows);
+
+    context->moveOrCopyResult(localResult, rows, result);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    // map(K1, V), function(K1, V) -> K2 -> map(K2, V)
+    return {exec::FunctionSignatureBuilder()
+                .typeVariable("K1")
+                .typeVariable("K2")
+                .typeVariable("V")
+                .returnType("map(K2,V)")
+                .argumentType("map(K1,V)")
+                .argumentType("function(K1,V,K2)")
+                .build()};
+  }
+
+ private:
+  void checkDuplicateKeys(
+      const MapVectorPtr& mapVector,
+      const SelectivityVector& rows) const {
+    static const char* kDuplicateKey = "Duplicate map keys are not allowed";
+
+    MapVector::canonicalize(mapVector);
+
+    auto offsets = mapVector->rawOffsets();
+    auto sizes = mapVector->rawSizes();
+    auto mapKeys = mapVector->mapKeys();
+    rows.applyToSelected([&](auto row) {
+      auto offset = offsets[row];
+      auto size = sizes[row];
+      for (auto i = 1; i < size; i++) {
+        if (mapKeys->equalValueAt(mapKeys.get(), offset + i, offset + i - 1)) {
+          VELOX_USER_FAIL("{}", kDuplicateKey);
```

Comment:
nit: Maybe also print out the value of the duplicate key?

Reply:
That's a good idea. Let me add this in a follow-up PR as it should be added to the map() function as well.

Reply:
@kagamiori #2260 adds duplicate key to the message.